src/Microsoft.AspNetCore.Razor.Evolution/ITagHelperBoundAttributeDescriptorBuilder.cs

```diff
@@ -0,0 +1,260 @@
+                {
+                    if (char.IsWhiteSpace(character) || InvalidNonWhitespaceAttributeNameCharacters.Contains(character))
+                    {
+                        var diagnostic = RazorDiagnosticFactory.CreateTagHelper_InvalidBoundAttributeName(
+                            _containingTypeName,
+                            _propertyName,
+                            _name,
+                            character);
+
+                        yield return diagnostic;
+                    }
+                }
+            }
+
+            if (_indexerNamePrefix != null)
+            {
+                if (_indexerNamePrefix.StartsWith(DataDashPrefix, StringComparison.OrdinalIgnoreCase))
+                {
+                    var diagnostic = RazorDiagnosticFactory.CreateTagHelper_InvalidBoundAttributePrefixStartsWith(
+                        _containingTypeName,
+                        _propertyName,
+                        _indexerNamePrefix);
+
+                    yield return diagnostic;
+                }
+                else if (_indexerNamePrefix.Length > 0 && string.IsNullOrWhiteSpace(_indexerNamePrefix))
+                {
+                    var diagnostic = RazorDiagnosticFactory.CreateTagHelper_InvalidBoundAttributeNullOrWhitespace(
+                        _containingTypeName,
+                        _propertyName);
+
+                    yield return diagnostic;
+                }
+                else
+                {
+                    foreach (var character in _indexerNamePrefix)
+                    {
+                        if (char.IsWhiteSpace(character) || InvalidNonWhitespaceAttributeNameCharacters.Contains(character))
+                        {
+                            var diagnostic = RazorDiagnosticFactory.CreateTagHelper_InvalidBoundAttributePrefix(
+                                _containingTypeName,
+                                _propertyName,
+                                _indexerNamePrefix,
+                                character);
+
+                            yield return diagnostic;
+                        }
+                    }
+                }
+            }
+        }
+
+        public BoundAttributeDescriptor Build()
+        {
+            if (!PrimitiveDisplayTypeNameLookups.TryGetValue(_typeName, out var simpleName))
+            {
+                simpleName = _typeName;
+            }
+
+            var displayName = $"{simpleName} {_containingTypeName}.{_propertyName}";
+            var descriptor = new ITagHelperBoundAttributeDescriptor(
+                _isEnum,
+                _name,
+                _propertyName,
+                _typeName,
+                _indexerNamePrefix,
+                _indexerValueTypeName,
+                _documentation,
+                displayName,
+                _propertyBag,
+                _diagnostics ?? Enumerable.Empty<RazorDiagnostic>());
+
+            return descriptor;
```

Comment:
Yes, because you want to be able to reuse the builder. Sorry, should have said shallow copy.